File: haskell/nucleotide-count/src/DNA.hs
module DNA (nucleotideCounts, Nucleotide(..)) where
import Data.Map (Map, fromList, alter, member)

data Nucleotide = A | G | C | T deriving (Eq, Ord, Show)

nucleotides :: [Nucleotide]
nucleotides = [A, G, C, T]

emptyNucleotideMap :: Map Nucleotide Int
emptyNucleotideMap = Data.Map.fromList $ map empty nucleotides
  where empty c = (c, 0)

nucleotideCounts :: String -> Either String (Map Nucleotide Int)
nucleotideCounts = nucleotideCountsAccum emptyNucleotideMap
  where nucleotideCountsAccum m "" = Right m
        nucleotideCountsAccum m (c:cs) =
          case fromChar c of
            Just n -> let m' = alter (fmap (+1)) n m
                      in nucleotideCountsAccum m' cs
            Nothing -> Left $ "Invalid nucleotide " ++ [c]

fromChar :: Char -> Maybe Nucleotide
fromChar 'A' = Just A
fromChar 'G' = Just G
fromChar 'C' = Just C
fromChar 'T' = Just T
fromChar _ = Nothing

File: kotlin/nucleotide-count/src/main/kotlin/Dna.kt
class Dna {
  val nucleotideCounts: Map<Char, Int>

  constructor(strand: String) {
    nucleotideCounts = getNucleotideCounts(strand)
  }

  private fun getNucleotideCounts(strand: String): Map<Char, Int> {
    val empty: MutableMap<Char, Int> = mutableMapOf('A' to 0, 'C' to 0, 'G' to 0, 'T' to 0)
    val increment: (MutableMap<Char, Int>, Char) -> MutableMap<Char, Int> = { acc, nucleotide ->
      val count = acc.get(nucleotide)
      if (count == null) {
        throw IllegalArgumentException("Invalid nucleotide $nucleotide")
      }
      acc.put(nucleotide, count + 1)
      acc
    }
    return strand.fold(empty, increment)
  }
}
